server/api/api/endpoints/nuclio.py

```diff
@@ -31,19 +37,103 @@ async def list_api_gateways(
         action=mlrun.common.schemas.AuthorizationAction.read,
         auth_info=auth_info,
     )
+    client = server.api.utils.clients.async_nuclio.Client(auth_info)
+    try:
+        api_gateways = await client.list_api_gateways(project)
+    finally:
+        await client.close_session()
```

Comment:
why do you need to explicitly close the session? that's some boilerplate here.
u may close the session on `__del__` functionality or, if you want to do it faster than python's GC then provide a `@contextlib.asynccontextmanager` decorator on top of the class that will close the session for you
e.g:

```python
async with server.api.utils.clients.async_nuclio.Client(auth_info) as client:
  api_gateways = await client.list_api_gateways(project)
```

while the class implements `__aenter__`
```python
class Client:
  def __init__(auth_info):
    self._auth_info = auth_info
  async def __aenter__():
    await self._session.initialize()
    yield self
  async def __aexit__(*args, **kwargs):
    await self._session.close()
```

(that's just an example of course

Reply:
@liranbg agree, reworked! Thanks for the detailed example :)